### PR TITLE
Course counts

### PR DESCRIPTION
**Pull request by janno42**
the new staff semester page does not have counts for the individual filter options like the course states had before. this pr adds these numbers.

### CODE REVIEW

File: evap/staff/views.py [hunk 1]
```diff
@@ -87,9 +87,61 @@ def __init__(self):
             self.first_start = datetime(9999, 1, 1)
             self.last_end = date(2000, 1, 1)
 
+    # course counts
```
Comment (He3lixxx) [inline evap/staff/views.py]: remove please :D

Edit: I like the rest, so I'll approve when this is changed :)

File: evap/staff/views.py [hunk 2]
```diff
@@ -87,9 +87,61 @@ def __init__(self):
...
+        if course.state in ['new', 'prepared', 'editor_approved', 'approved', 'in_evaluation', 'evaluated']:
+            course_counts.num_evaluation_not_finished += 1
+        elif course.state == 'reviewed':
+            if course.is_single_result or course.final_grade_documents or course.gets_no_grade_documents:
+                course_counts.num_not_published_warning += 1
+            else:
+                course_counts.num_not_published += 1
+        if course.state == 'published':
+            course_counts.num_published += 1
+
```
[the remark this answers is not in the document]
Comment (karyon) [reply evap/staff/views.py]: maybe you can simply iterate over all the buttons and add that span for each after applying datatables.

**Comment by janno42** [comment]
now the counts are added via javascript. also the selected state is restored after page reload.

File: evap/staff/templates/staff_semester_view.html [hunk 3]
```diff
@@ -292,50 +359,49 @@ <h3>
...
+                    };
+                    // add counts to each filter button
+                    $("#course-filter-buttons [data-filter]").each(function(i, obj) {
+                        var count = countIcons(obj);
+                        if (count > 0) {
+                            $('<span class="badge badge-pill badge-btn">' + count + '</span>').appendTo(obj);
+                        }
+                        else {
+                            $('<span class="badge badge-pill badge-btn-zero">' + count + '</span>').appendTo(obj);
+                        }
```
Comment (karyon) [inline evap/staff/templates/staff_semester_view.html]: so i just learned about [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals), how about this:

``$(`<span class="badge badge-pill ${count == 0 ? 'badge-btn-zero' : 'badge-btn'}> ${count} </span>`).appendTo(obj);``

File: evap/staff/templates/staff_semester_view.html [hunk 4]
```diff
@@ -292,50 +359,49 @@ <h3>
...
+                        localStorage.removeItem('course-filter_{{ semester.id }}');
+                        if($("[data-filter].active").length)
+                        {
+                            localStorage['course-filter_{{ semester.id }}'] = $("[data-filter].active")[0].dataset.filter;
+                        }
+                    });
+                    // helper function to count number of lines with certain icon
+                    countIcons = function(obj) {
+                        var count = 0;
+                        $.each([1, 2, 3, 4], function(index, value) {
```
Comment (karyon) [inline evap/staff/templates/staff_semester_view.html]: you could use forEach and the arrow notation for something like this:
```
[1,2,3,4].forEach( i => { 
    count += table.column(i).data().filter( (value, index) => value.includes(obj.dataset.filter)).count();
}
```

okay fancy one-liner incoming
`return [1,2,3,4].reduce( (sum, i) => sum + table.column(i).data().filter( (value, index) => value.includes(obj.dataset.filter)).count(), 0);`

:) but yeah, that probably doesn't help anymore...

in any case, i would inline the function (and maybe call the variable iconCount).

File: evap/staff/templates/staff_semester_view.html [hunk 5]
```diff
@@ -292,50 +359,49 @@ <h3>
...
+                        $("[data-filter].active").not(button).removeClass("active");
                         table.draw();
-                    } );
-                    activateTooltips('.filter-toolbar');
-                },
+
+                        // store which button was selected
+                        localStorage.removeItem('course-filter_{{ semester.id }}');
+                        if($("[data-filter].active").length)
+                        {
```
Comment (karyon) [inline evap/staff/templates/staff_semester_view.html]: brace on the same line please

File: evap/staff/templates/staff_semester_view.html [hunk 6]
```diff
@@ -292,50 +359,49 @@ <h3>
...
+                    // add counts to each filter button
+                    $("#course-filter-buttons [data-filter]").each(function(i, obj) {
+                        var count = countIcons(obj);
+                        if (count > 0) {
+                            $('<span class="badge badge-pill badge-btn">' + count + '</span>').appendTo(obj);
+                        }
+                        else {
+                            $('<span class="badge badge-pill badge-btn-zero">' + count + '</span>').appendTo(obj);
+                        }
+                    });
```
Comment (karyon) [inline evap/staff/templates/staff_semester_view.html]: maybe swap this with the `.click` part above so the localstorage stuff is located together?

File: evap/staff/templates/staff_semester_view.html (hunk 3 above)
Comment (janno42) [reply evap/staff/templates/staff_semester_view.html]: `${count}` works fine. the inline if does work on it's own but when adding it like in your snippet, nothing happens at all.

**Comment by karyon** [comment]
Random guess, die you use backticks as required for template literals?

**Comment by janno42** [comment]
yes, i am now using template literals, just the `if ? :` part does not work

**Comment by karyon** [comment]
``$(`<span class="badge badge-pill ${iconCount == 0 ? 'badge-btn-zero' : 'badge-btn'}">${iconCount}</span>`).appendTo(obj);`` works.

**Pull request merged**